### PR TITLE
Add voms-proxy-direct tests (SOFTWARE-4883)

### PR DESCRIPTION
**Pull request by brianhlin**
I think I based this on the right branch. Kicked off and waiting on test results.

### CODE REVIEW

File: osgtest/library/voms.py [hunk 1]
```diff
@@ -116,6 +117,59 @@ def destroy_voms_conf(vo):
...
+        cert_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/usercert.pem')
+    if not key_path:
+        key_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/userkey.pem')
+
+    # voms-proxy-direct requires that the user cert/key are owned by the user running the command
+    filemap = dict()
+    for src in (cert_path, key_path):
+        dest = tempfile.NamedTemporaryFile()
+        filemap[src] = dest.name
+        shutil.copy2(src, dest.name)
```
Comment (matyasselmeci) [inline osgtest/library/voms.py:153]: According to the help, shutil.copy2 copies the permissions from the original file so you will need to chown the files to root after copying.

File: osgtest/library/voms.py [hunk 2]
```diff
@@ -116,6 +117,59 @@ def destroy_voms_conf(vo):
...
+    if not cert_path:
+        cert_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/usercert.pem')
+    if not key_path:
+        key_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/userkey.pem')
+
+    # voms-proxy-direct requires that the user cert/key are owned by the user running the command
+    filemap = dict()
+    for src in (cert_path, key_path):
+        dest = tempfile.NamedTemporaryFile()
+        filemap[src] = dest.name
```
Comment (matyasselmeci) [inline osgtest/library/voms.py]: This adds a string to filemap, not the NamedTemporaryFile, which is why the close() call at line 170 barfs.

File: osgtest/library/voms.py [hunk 3]
```diff
@@ -116,6 +117,59 @@ def destroy_voms_conf(vo):
...
+                   '-voms', VONAME,
+                   '-uri', core.get_hostname(),
+                   '-fqan', fqan,
+                   '-cert', filemap[cert_path],
+                   '-key', filemap[key_path],
+                   '-hostcert', core.config['certs.vomscert'],
+                   '-hostkey', core.config['certs.vomskey'],
+                   '-out', proxy_path)
+        core.check_system(command, 'Run voms-proxy-direct', stdin=password)
+        os.chown(username, uid, gid)
```
Comment (matyasselmeci) [inline osgtest/library/voms.py]:
```suggestion
        os.chown(proxy_path, uid, gid)
```

File: osgtest/library/voms.py (hunk 1 above)
Comment (brianhlin) [reply osgtest/library/voms.py:153]: I think we're in the clear:

> Warning Even the higher-level file copying functions (shutil.copy(), shutil.copy2()) cannot copy all file metadata.
>
> On POSIX platforms, this means that file owner and group are lost as well as ACLs. On Mac OS, the resource fork and other metadata are not used. This means that resources will be lost and file type and creator codes will not be correct. On Windows, file owners, ACLs and alternate data streams are not copied.

File: osgtest/library/voms.py (hunk 2 above)
Comment (brianhlin) [reply osgtest/library/voms.py]: Good catch, fixed

File: osgtest/library/voms.py (hunk 1 above)
Comment (matyasselmeci) [reply osgtest/library/voms.py:153]: Just tested it and it looks like you're right: even as root, ownership isn't preserved. Seems like an odd thing to omit though I guess this time it's in our favor.

**Review by matyasselmeci — APPROVED** [review]
LGTM (assuming tests pass this time)

**Comment by matyasselmeci** [comment]
Note that closing the NamedTemporaryFile isn't necessary: it ill be deleted once the object goes out of scope (so at the end of the function).

File: osgtest/library/voms.py [hunk 4]
```diff
@@ -116,6 +117,55 @@ def destroy_voms_conf(vo):
...
+        user = pwd.getpwnam(username)
+        uid = user.pw_uid
+        gid = user.pw_gid
+    else:
+        username = core.options.username
+        uid = core.state['user.uid']
+        gid = core.state['user.gid']
+
+    if not password:
+        password = core.options.password + '\n'
```
Comment (matyasselmeci) [inline osgtest/library/voms.py]:
```suggestion
    if not password:
        password = core.options.password
    if not password.endswith('\n'):
        password = password + '\n'
```

**Comment by brianhlin** [comment]
Test results look good: https://osg-sw-submit.chtc.wisc.edu/tests/20211109-1048/results.html

**Pull request merged**